OOM failure slipping through

**Summary**
It looks like a job in the gbs-uber-keyfiles pipeline failed with Out of Memory error, redun reported no failed jobs and no outstanding jobs.

**Details**
What prompted me to investigate was that mandatory output files were missing from the KGD results folder.
The only files present are:

```
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox 450757 Sep  3 13:15 SampleStatsRaw.csv
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox 279087 Sep  3 13:15 seqID.csv
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox  45682 Sep  3 13:15 AlleleFreq.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox  14935 Sep  3 13:16 CallRate.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox  14233 Sep  3 13:16 SNPCallRate.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox 619569 Sep  3 13:16 SampleStats.csv
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox   6797 Sep  3 13:16 SampDepth.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox   7307 Sep  3 13:16 SampDepth-scored.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox  13341 Sep  3 13:16 SampDepthHist.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox   9177 Sep  3 13:16 SampDepthCR.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox  12068 Sep  3 13:16 SNPDepthHist.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox  17782 Sep  3 13:16 SNPDepth.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox 448072 Sep  3 13:16 finplot.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox 226510 Sep  3 13:16 HWdisMAFsig.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox  11584 Sep  3 13:16 X2star-QQ.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox   9272 Sep  3 13:16 LRT-QQ.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox   7317 Sep  3 13:16 LRT-hist.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox   7523 Sep  3 13:16 MAF.png
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox  92104 Sep  3 13:16 SampleStatsRawCombined.csv
-rw-rw-r--. 1 brauningr@agresearch.co.nz 2023-rb-sandbox   6719 Sep  3 13:17 Co-call-.png

```

`KGD.stderr` does not report any problems.
`KGD.stdout` finishes with

```
Analysing 9591 individuals and 24172 SNPs
Proportion of missing genotypes:  0.614476 Callrate: 0.385524
Mean sample depth: 1.347506
Control status of removed samples

     <NA>
  77    0
Multiple plates - plate plots not produced
Calculating G matrix, analysis code:
Using global allele frequencies
# SNPs:  24172
# individuals:  9591
Mean co-call rate (for sample pairs): 0.1884546
Min  co-call rate (for sample pairs): 8.274036e-05
Proportion of missing genotypes:  0.614476 Callrate: 0.385524
Mean sample depth: 1.347506
```
That means it probably died when running `calcG`

`seffs` did not list `uber-keyfiles.KGD`
`gbs-uber-keyfiles.log` contained some clues:

```
2025-09-03 13:24 redun        INFO
2025-09-03 13:24 redun        INFO     | JOB STATUS 2025/09/03 13:24:37
2025-09-03 13:24 redun        INFO     | TASK                          PENDING RUNNING  FAILED  CACHED    DONE   TOTAL
2025-09-03 13:24 redun        INFO     |
2025-09-03 13:24 redun        INFO     | ALL                                 1       2       0      23       2      28
2025-09-03 13:24 redun        INFO     | _merge_results_and_counts           0       0       0       1       0       1
2025-09-03 13:24 redun        INFO     | agr.gbs_uber_keyfiles.main          0       1       0       0       0       1
2025-09-03 13:24 redun        INFO     | create_consolidated_tag_count       0       0       0       1       0       1
2025-09-03 13:24 redun        INFO     | demultiplex                         0       0       0       1       0       1
2025-09-03 13:24 redun        INFO     | get_fastq_to_tag_count              0       0       0      17       0      17
2025-09-03 13:24 redun        INFO     | gusbase                             1       0       0       0       0       1
2025-09-03 13:24 redun        INFO     | kgd                                 0       1       0       0       0       1
2025-09-03 13:24 redun        INFO     | map_info_to_hap_map                 0       0       0       0       1       1
2025-09-03 13:24 redun        INFO     | merge_taxa_tag_count                0       0       0       1       0       1
2025-09-03 13:24 redun        INFO     | tag_count_to_tag_pair               0       0       0       1       0       1
2025-09-03 13:24 redun        INFO     | tag_pair_to_tbt                     0       0       0       1       0       1
2025-09-03 13:24 redun        INFO     | tbt_to_map_info                     0       0       0       0       1       1
2025-09-03 13:24 redun        INFO
2025-09-03 13:24 redun        INFO
2025-09-03 13:24 psij.executors.batch.batch_scheduler_executor INFO     Polling for 1 jobs
2025-09-03 13:24 psij.executors.batch.batch_scheduler_executor DEBUG    Output from status command: JOBID               ST                  REASON
2461241             OOM                 OutOfMemory

2025-09-03 13:24 psij.executors.batch.batch_scheduler_executor WARNING  Polling error: Failed to poll for job status: Traceback (most recent call last):
  File "/nix/store/pymj9h1ww464vjh9d59nzl4bwsqykprz-python3.12-psij/lib/python3.12/site-packages/psij/executors/batch/batch_scheduler_executor.py", line 677, in _poll
    status_map = self.executor.parse_status_output(exit_code, out)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/pymj9h1ww464vjh9d59nzl4bwsqykprz-python3.12-psij/lib/python3.12/site-packages/psij/executors/batch/slurm.py", line 169, in parse_status_output
    msg = self._get_message(cols[2]) if state == JobState.FAILED else None
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/pymj9h1ww464vjh9d59nzl4bwsqykprz-python3.12-psij/lib/python3.12/site-packages/psij/executors/batch/slurm.py", line 183, in _get_message
    assert reason in SlurmJobExecutor._REASONS_MAP
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

With the PID I can see

```
               JobID              JobName  Partition    Account  AllocCPUS                State ExitCode  MaxVMSize
-------------------- -------------------- ---------- ---------- ---------- -------------------- -------- ----------
             2461241    uber-keyfiles.KGD    compute 2023-rb-s+          4        OUT_OF_MEMORY    0:125
       2461241.batch                batch            2023-rb-s+          4        OUT_OF_MEMORY    0:125     34.62G
      2461241.extern               extern            2023-rb-s+          4            COMPLETED      0:0      0.15G
```

I launched the pipeline from `/home/agresearch.co.nz/brauningr/vc/agr_github/gbs-uber-keyfiles/` with `redun run pipeline.py main --enzyme PstI-MspI --keyfile /scratch/2023-rb-sandbox/uber_keyfile_workdir/btprawn_20250731/PstI-MspI/sample_info.key --workdir /scratch/2023-rb-sandbox/uber_keyfile_workdir/btprawn_20250805`
Log file `/home/agresearch.co.nz/brauningr/vc/agr_github/gbs-uber-keyfiles/gbs-uber-keyfiles.log`
Outputs at `/scratch/2023-rb-sandbox/uber_keyfile_workdir/btprawn_20250805/`

I instrumented the assert statement in psij-python, and found this.  It's a bug in psij-python. I will make a fix, test it, and contribute a PR upstream.
```
2025-11-12 13:46 redun        INFO
2025-11-12 13:46 redun        INFO     | JOB STATUS 2025/11/12 13:46:10
2025-11-12 13:46 redun        INFO     | TASK               PENDING RUNNING  FAILED  CACHED    DONE   TOTAL
2025-11-12 13:46 redun        INFO     |
2025-11-12 13:46 redun        INFO     | ALL                      0       2       0       0       0       2
2025-11-12 13:46 redun        INFO     | psij.oom.main            0       1       0       0       0       1
2025-11-12 13:46 redun        INFO     | psij.oom.memtester       0       1       0       0       0       1
2025-11-12 13:46 redun        INFO
2025-11-12 13:46 redun        INFO
2025-11-12 13:46 psij.executors.batch.batch_scheduler_executor INFO     Polling for 1 jobs
2025-11-12 13:46 psij.executors.batch.batch_scheduler_executor DEBUG    Output from status command: JOBID               ST                  REASON
2791176             OOM                 OutOfMemory

2025-11-12 13:46 psij.executors.batch.batch_scheduler_executor WARNING  Polling error: Failed to poll for job status: Traceback (most recent call last):
  File "/nix/store/f9x5y8yjyy6iwjal6c0z256hpzb7lv9r-python3-3.12.10-env/lib/python3.12/site-packages/psij/executors/batch/batch_scheduler_executor.py", line 698, in _poll
    status_map = exec.parse_status_output(exit_code, out)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/f9x5y8yjyy6iwjal6c0z256hpzb7lv9r-python3-3.12.10-env/lib/python3.12/site-packages/psij/executors/batch/slurm.py", line 169, in parse_status_output
    msg = self._get_message(cols[2]) if state == JobState.FAILED else None
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/f9x5y8yjyy6iwjal6c0z256hpzb7lv9r-python3-3.12.10-env/lib/python3.12/site-packages/psij/executors/batch/slurm.py", line 184, in _get_message
    reason in SlurmJobExecutor._REASONS_MAP
AssertionError: reason 'OutOfMemory' not in reasons map
```

--- a/flake.nix
+++ b/flake.nix
@@ -42,7 +42,10 @@
       inputs.nixpkgs.follows = "nixpkgs";
     };
     redun-psij = {
-      url = "github:AgResearch/redun_psij/main";
+      # TODO revert to main branch once this is merged and referenced:
+      # https://github.com/ExaWorks/psij-python/pull/534
+      # url = "github:AgResearch/redun_psij/main";
+      url = "github:AgResearch/redun_psij?ref=refs/heads/slurm-oom-reason";
       inputs = {
         nixpkgs.follows = "nixpkgs";
         redun.follows = "redun";

